File: src/services/deepgram/tts.js
import { createClient } from '@deepgram/sdk';
import { config } from '../../config/config.js';

/**
 * Deepgram Text-to-Speech (TTS) Service
 * 
 * Converts text responses from the LLM into natural-sounding speech audio.
 * Returns raw audio buffers that can be streamed via WebSocket to the client.
 */
class DeepgramTTS {
  constructor() {
    this.client = createClient(config.deepgramApiKey);
  }

  /**
   * Convert text to speech audio
   * @param {string} text - The text to convert to speech
   * @returns {Promise<Buffer>} - Raw audio buffer (linear16, 24kHz)
   */
  async synthesize(text) {
    if (!text || text.trim() === '') {
      throw new Error('TTS: Empty text provided');
    }

    try {
      const response = await this.client.speak.request(
        { text },
        {
          model: config.tts.model,
          encoding: config.tts.encoding,
          sample_rate: config.tts.sample_rate,
          container: config.tts.container,
        }
      );

      const stream = await response.getStream();
      if (!stream) {
        throw new Error('TTS: No audio stream returned');
      }

      // Collect all audio chunks into a single buffer
      const chunks = [];
      const reader = stream.getReader();

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        chunks.push(value);
      }

      const audioBuffer = Buffer.concat(chunks.map(c => Buffer.from(c)));
      console.log(`🔊 [TTS] Generated ${audioBuffer.length} bytes of audio`);
      return audioBuffer;
    } catch (error) {
      console.error('❌ [TTS] Synthesis failed:', error.message);
      throw error;
    }
  }

  /**
   * Convert text to speech and stream audio chunks via callback
   * Lower latency than synthesize() since audio starts playing before
   * the full response is generated.
   * 
   * @param {string} text - Text to convert
   * @param {Function} onChunk - Called with each audio chunk (Buffer)
   * @param {Function} onDone - Called when streaming is complete
   */
  async streamSynthesize(text, onChunk, onDone) {
    if (!text || text.trim() === '') {
      if (onDone) onDone();
      return;
    }

    try {
      const response = await this.client.speak.request(
        { text },
        {
          model: config.tts.model,
          encoding: config.tts.encoding,
          sample_rate: config.tts.sample_rate,
          container: config.tts.container,
        }
      );

      const stream = await response.getStream();
      if (!stream) {
        throw new Error('TTS: No audio stream returned');
      }

      const reader = stream.getReader();

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        const chunk = Buffer.from(value);
        if (onChunk) onChunk(chunk);
      }

      console.log('🔊 [TTS] Stream complete');
      if (onDone) onDone();
    } catch (error) {
      console.error('❌ [TTS] Stream synthesis failed:', error.message);
      throw error;
    }
  }
}

export default DeepgramTTS;

File: app/services/deepgram/tts.py
"""
Deepgram Text-to-Speech (TTS) Service

Converts text responses from the LLM into natural-sounding speech audio.
Returns raw audio buffers that can be streamed via WebSocket to the client.
Includes automatic retry with exponential backoff for resilience.
"""

import asyncio
import logging
import random
from typing import Callable, Optional

from deepgram import DeepgramClient, SpeakOptions

from app.config import config

logger = logging.getLogger(__name__)


# ─── Retry helper (inlined) ─────────────────────────────────

async def _with_retry(fn, max_retries=3, initial_delay_ms=300,
                      max_delay_ms=2000, backoff_multiplier=2, name="TTS"):
    """Execute *fn* with exponential-backoff retry."""
    last_err = None
    for attempt in range(max_retries + 1):
        try:
            return await fn()
        except Exception as exc:
            last_err = exc
            if attempt == max_retries:
                logger.error("❌ [%s] Failed after %d attempts: %s", name, max_retries + 1, exc)
                raise
            delay = min(initial_delay_ms * (backoff_multiplier ** attempt), max_delay_ms)
            jitter = delay * 0.2 * (random.random() * 2 - 1)
            wait = (delay + jitter) / 1000.0
            logger.warning("⚠️ [%s] Attempt %d failed: %s. Retrying in %.2fs…",
                           name, attempt + 1, exc, wait)
            await asyncio.sleep(wait)
    raise last_err  # type: ignore[misc]


class DeepgramTTS:
    def __init__(self):
        self.client = DeepgramClient(config["deepgram_api_key"])

    async def synthesize(self, text: str) -> bytes:
        """
        Convert text to speech audio (with automatic retry).

        Args:
            text: The text to convert to speech.

        Returns:
            Raw audio bytes (linear16, 24 kHz).
        """
        if not text or text.strip() == "":
            raise ValueError("TTS: Empty text provided")

        async def _do():
            tts_cfg = config["tts"]
            options = SpeakOptions(
                model=tts_cfg["model"],
                encoding=tts_cfg["encoding"],
                sample_rate=tts_cfg["sample_rate"],
                container=tts_cfg["container"],
            )
            response = await self.client.speak.asyncrest.v("1").stream_memory(
                {"text": text}, options
            )
            audio_data = response.stream.read()
            logger.info("🔊 [TTS] Generated %d bytes of audio", len(audio_data))
            return audio_data

        return await _with_retry(_do, name="TTS")

    async def stream_synthesize(
        self,
        text: str,
        on_chunk: Optional[Callable[[bytes], None]] = None,
        on_done: Optional[Callable[[], None]] = None,
    ):
        """
        Convert text to speech and stream audio chunks via callback.
        Lower latency than synthesize() since audio starts playing before
        the full response is generated.

        Args:
            text: Text to convert.
            on_chunk: Called with each audio chunk (bytes).
            on_done: Called when streaming is complete.
        """
        if not text or text.strip() == "":
            if on_done:
                on_done()
            return

        async def _do():
            tts_cfg = config["tts"]
            options = SpeakOptions(
                model=tts_cfg["model"],
                encoding=tts_cfg["encoding"],
                sample_rate=tts_cfg["sample_rate"],
                container=tts_cfg["container"],
            )
            response = await self.client.speak.asyncrest.v("1").stream_memory(
                {"text": text}, options
            )
            # Read in chunks
            chunk_size = 4096
            while True:
                chunk = response.stream.read(chunk_size)
                if not chunk:
                    break
                if on_chunk:
                    on_chunk(chunk)

            logger.info("🔊 [TTS] Stream complete")
            if on_done:
                on_done()

        await _with_retry(_do, name="TTS-Stream")
